### ISSUE TITLE
rustc does not warn in a safe fn when a trivially unsafe operation is used on an argument

### ISSUE THREAD

**Issue by ds84182** [rust-lang/rust#162090]
### Code

```Rust
fn oopsie_woopsie_uwu(ptr: *mut u32) {
    unsafe { ptr.write(123); }
}

fn main() {
    oopsie_woopsie_uwu(0 as _);
}
```

### Current output

```Shell
No diagnostics
```

### Desired output

```Shell
warning: call to function `std::ptr::mut_ptr::<impl *mut T>::write` is unsafe but the pointer is an argument of a safe function
 --> src/main.rs:2:7
  |
1 | fn oopsie_woopsie_uwu(ptr: *mut u32) {
  |                       ^^^^^^^^^^^^^
  |                       argument declared here
  |
2 |     { ptr.write(123); }
  |       ^^^^^^^^^^^^^^ call to unsafe function
  |
  = note: consult the function's documentation for information on how to avoid undefined behavior
```

### Rationale and extra context

I propose a simple lint warning the developer that the function should be marked unsafe, otherwise preconditions are easily ignored. For now this should only include integers (`unchecked_*` functions) and pointers (read, write, arithmetic). It should not include struct fields unless the field is marked as `pub` and is mutable.

It could also include FFI functions that accept a pointer but are marked as safe.

### Other cases

```Rust

```

### Rust Version

```Shell
1.98.0 (playground)
```

**Comment by joshtriplett** [comment]
In general, it's reasonable for safe functions to wrap unsafe operations. However, I do think there's merit to the idea of linting against *obvious* "safe version of standard unsafe operation" cases, like "take a pointer and write to it, with no validation".

**Comment by cyrgani** [comment]
Notably, there is already a similar clippy lint: [`not_unsafe_ptr_arg_deref`](https://rust-lang.github.io/rust-clippy/main/index.html#not_unsafe_ptr_arg_deref), though it only lints public functions at the moment.

**Comment by rperier** [comment]
Do you think it's better that I add this into `rustc_lint` or into  `rustc_hir_typeck` ?
